File: CreatePQTFile.cs
#r "System.IO"
#r "System.IO.Compression.FileSystem"
#r "System.Xml.Linq"

using System.IO;
using System.IO.Compression;
using System.Xml.Linq;

public class QueryMetadata
{
    public string QueryName { get; set; }
    public object QueryGroupId { get; set; }
    public string LastKnownIsParameter { get; set; }
    public string LastKnownResultTypeName { get; set; }
    public bool LoadEnabled { get; set; }
    public bool IsHidden { get; set; }
}
public class RootObject
{
    public string DocumentLocale { get; set; }
    public string EngineVersion { get; set; }
    public List<QueryMetadata> QueriesMetadata { get; set; }
    public List<object> QueryGroups { get; set; }
}

// ***Make sure to fill in the folderPath parameter***

string fileName = "PowerQueryTemplate"; // This is the name of the .pqt file which will be created
string folderPath = @"C:\Desktop\MyFolder"; // This is the folder where the .pqt file will be created
string subFolderPath = folderPath + @"\" + "pqtnewfolder";
Directory.CreateDirectory(subFolderPath);

// STEP 1: Create MashupDocument.pq
string mdfileName = "MashupDocument.pq";
string mdfilePath = subFolderPath + @"\" + mdfileName;
var sb = new System.Text.StringBuilder();
string newline = Environment.NewLine;
sb.Append("section Section1;");
foreach (var t in Model.Tables)
{
    string tName = "#\"" + t.Name + "\"";
    sb.Append(newline + "shared " + tName + " = ");
    foreach (var p in t.Partitions)
    {
        sb.Append(p.Query + ";");
    }
}

foreach (var e in Model.Expressions)
{
    string expr = e.Expression;
    string eName = "#\"" + e.Name + "\"";
    sb.Append(newline + "shared " + eName + " = " + expr + ";");
}

File.WriteAllText(mdfilePath, sb.ToString());

// STEP 2: Create the MashupMetadata.json file based on looping through the table names to create a QueryMetadata object for each table
string mmFileName = "MashupMetadata.json";
string mmFilePath = subFolderPath + @"\" + mmFileName;

var queryMetadata = new List<QueryMetadata>();

// Create a QueryMetadata object for each table
foreach (var t in Model.Tables)
{
    string tName = t.Name;
    queryMetadata.Add( new QueryMetadata {QueryName = tName, QueryGroupId = null, LastKnownIsParameter = null, LastKnownResultTypeName = null, LoadEnabled = true, IsHidden = false});
}
foreach (var e in Model.Expressions)
{
    string eName = e.Name;
    if (e.Kind.ToString() == "M")
    {
        queryMetadata.Add( new QueryMetadata {QueryName = eName, QueryGroupId = null, LastKnownIsParameter = null, LastKnownResultTypeName = null, LoadEnabled = true, IsHidden = false});
    }
    else
    {
        queryMetadata.Add( new QueryMetadata {QueryName = eName, QueryGroupId = null, LastKnownIsParameter = null, LastKnownResultTypeName = null, LoadEnabled = false, IsHidden = false});
    }
}

// Create a single Root object
var rootObject = new {DocumentLocale = "en-US", EngineVersion = "2.126.453.0", QueriesMetadata = queryMetadata, QueryGroups = new List<object>()};
string jsonContent = JsonConvert.SerializeObject(rootObject, Formatting.Indented);
System.IO.File.WriteAllText(mmFilePath, jsonContent);

// STEP 3: Create Metadata.json file
string mFileName = "Metadata.json";
string mFilePath = subFolderPath + @"\" + mFileName;
var metaData = new {Name = fileName, Description = "", Version = "1.0.0.0"};
jsonContent = JsonConvert.SerializeObject(metaData, Formatting.Indented);
System.IO.File.WriteAllText(mFilePath, jsonContent);

// STEP 4: Create [Content_Types].xml file
XNamespace ns = "http://schemas.openxmlformats.org/package/2006/content-types";

XDocument xmlDocument = new XDocument(
    new XDeclaration("1.0", "utf-8", null),
    new XElement(ns + "Types",
        new XElement(ns + "Default",
            new XAttribute("Extension", "json"),
            new XAttribute("ContentType", "application/json")),
        new XElement(ns + "Default",
            new XAttribute("Extension", "pq"),
            new XAttribute("ContentType", "application/x-ms-m"))
    )
);

string xmlFileName = "[Content_Types].xml";
string xmlFilePath = subFolderPath + @"\" + xmlFileName;
xmlDocument.Save(xmlFilePath);

// STEP 5: Zip up the 4 files
string zipFileName = fileName + ".zip";
string zipFilePath = folderPath + @"\" + zipFileName;
ZipFile.CreateFromDirectory(subFolderPath, zipFilePath);

// STEP 6: Convert the zip file back into a pqt file
string newExt = ".pqt";
string directory = Path.GetDirectoryName(zipFilePath);
string fileNameWithoutExtension = Path.GetFileNameWithoutExtension(zipFilePath);
string newFilePath = Path.Combine(directory, fileNameWithoutExtension + newExt);
File.Move(zipFilePath, newFilePath);

// STEP 7: Delete subFolder directory which is no longer needed
Directory.Delete(subFolderPath, true);

Info("The " + fileName + ".pqt file has been created here: " + folderPath);

File: CreatePQTFile.py
import sempy
import sempy.fabric as fabric
import pandas as pd
import json
import os
import xml.etree.ElementTree as ET
import shutil
from sempy.fabric._client import DatasetXmlaClient
from sempy.fabric._cache import _get_or_create_workspace_client
sempy.fabric._client._utils._init_analysis_services()

def create_pqt_file(datasetName, fileName = 'PowerQueryTemplate'):

    fileName = 'PowerQueryTemplate'
    folderPath = '/lakehouse/default/Files'
    subFolderPath = os.path.join(folderPath, 'pqtnewfolder')
    os.makedirs(subFolderPath, exist_ok=True)

    def list_tables(datasetName, workspaceName = None):

        if workspaceName == None:
            workspaceId = fabric.get_workspace_id()
            workspaceName = fabric.resolve_workspace_name(workspaceId)

        workspace_client = _get_or_create_workspace_client(workspaceName)
        ds = workspace_client.get_dataset(datasetName)
        m = ds.Model

        header = pd.DataFrame(columns=['Name', 'Refresh Policy', 'Source Expression'])
        df = pd.DataFrame(header)

        for t in m.Tables:
            rPolicy = bool(t.RefreshPolicy)
            sourceExpression = None

            if rPolicy:
                sourceExpression = t.RefreshPolicy.SourceExpression

            new_data = {'Name': t.Name, 'Refresh Policy': rPolicy, 'Source Expression': sourceExpression}
            df = pd.concat([df, pd.DataFrame(new_data, index=[0])], ignore_index=True)

        return df

    dfP = fabric.list_partitions(datasetName)
    dfT = list_tables(datasetName)
    dfE = fabric.list_expressions(datasetName)

    # Check if M-partitions are used
    if any(dfP['Source Type'] == 'M'):
        class QueryMetadata:
            def __init__(self, QueryName, QueryGroupId=None, LastKnownIsParameter=None, LastKnownResultTypeName=None, LoadEnabled=True, IsHidden=False):
                self.QueryName = QueryName
                self.QueryGroupId = QueryGroupId
                self.LastKnownIsParameter = LastKnownIsParameter
                self.LastKnownResultTypeName = LastKnownResultTypeName
                self.LoadEnabled = LoadEnabled
                self.IsHidden = IsHidden

        class RootObject:
            def __init__(self, DocumentLocale, EngineVersion, QueriesMetadata, QueryGroups=None):
                if QueryGroups is None:
                    QueryGroups = []
                self.DocumentLocale = DocumentLocale
                self.EngineVersion = EngineVersion
                self.QueriesMetadata = QueriesMetadata
                self.QueryGroups = QueryGroups

        # STEP 1: Create MashupDocument.pq
        mdfileName = 'MashupDocument.pq'
        mdFilePath = os.path.join(subFolderPath, mdfileName)
        sb = 'section Section1;'
        for table_name in dfP['Table Name'].unique():
            tName = '#\"' + table_name + '"'
            sb = sb + '\n' + 'shared ' + tName + ' = '
            sourceExpression = dfT.loc[(dfT['Name'] == table_name), 'Source Expression'].iloc[0]
            refreshPolicy = dfT.loc[(dfT['Name'] == table_name), 'Refresh Policy'].iloc[0]

            if refreshPolicy == True:
                sb = sb + sourceExpression + ';'

            partitions_in_table = dfP.loc[dfP['Table Name'] == table_name, 'Partition Name'].unique()

            i=1
            for partition_name in partitions_in_table:
                pSourceType = dfP.loc[(dfP['Table Name'] == table_name) & (dfP['Partition Name'] == partition_name), 'Source Type'].iloc[0]
                if refreshPolicy == False and pSourceType == 'M' and i==1:
                    pQuery = dfP.loc[(dfP['Table Name'] == table_name) & (dfP['Partition Name'] == partition_name), 'Query'].iloc[0]
                    sb = sb + pQuery + ';'
                i+=1

        for index, row in dfE.iterrows():
            expr = row['Expression']
            eName = row['Name']
            eName = '#"' + eName + '"'
            sb = sb + '\n' + "shared " + eName + " = " + expr + ";"

        with open(mdFilePath, 'w') as file:
            file.write(sb)

        # STEP 2: Create the MashupMetadata.json file
        mmfileName = 'MashupMetadata.json'
        mmFilePath = os.path.join(subFolderPath, mmfileName)
        queryMetadata = []

        for i, r in dfT.iterrows():
            tName = r['Name']
            queryMetadata.append(QueryMetadata(tName, None, None, None, True, False))

        for i, r in dfE.iterrows():
            eName = r['Name']
            eKind = r['Kind']
            if eKind == 'M':
                queryMetadata.append(QueryMetadata(eName, None, None, None, True, False))
            else:
                queryMetadata.append(QueryMetadata(eName, None, None, None, False, False))

        rootObject = RootObject("en-US", "2.126.453.0", queryMetadata)

        def obj_to_dict(obj):
            if isinstance(obj, list):
                return [obj_to_dict(e) for e in obj]
            elif hasattr(obj, "__dict__"):
                return {k: obj_to_dict(v) for k, v in obj.__dict__.items()}
            else:
                return obj
        jsonContent = json.dumps(obj_to_dict(rootObject), indent=4)

        with open(mmFilePath, 'w') as json_file:
            json_file.write(jsonContent)

        # STEP 3: Create Metadata.json file
        mFileName = 'Metadata.json'
        mFilePath = os.path.join(subFolderPath, mFileName)
        metaData = {"Name": "fileName", "Description": "", "Version": "1.0.0.0"}
        jsonContent = json.dumps(metaData, indent=4)

        with open(mFilePath, 'w') as json_file:
            json_file.write(jsonContent)

        # STEP 4: Create [Content_Types].xml file:
        ns = 'http://schemas.openxmlformats.org/package/2006/content-types'
        ET.register_namespace('', ns) 
        types = ET.Element("{%s}Types" % ns)
        default1 = ET.SubElement(types, "{%s}Default" % ns, {"Extension": "json", "ContentType": "application/json"})
        default2 = ET.SubElement(types, "{%s}Default" % ns, {"Extension": "pq", "ContentType": "application/x-ms-m"})
        xmlDocument = ET.ElementTree(types)
        xmlFileName = '[Content_Types].xml'
        xmlFilePath = os.path.join(subFolderPath, xmlFileName)
        xmlDocument.write(xmlFilePath, xml_declaration=True, encoding='utf-8', method='xml')

        # STEP 5: Zip up the 4 files
        zipFileName = fileName + '.zip'
        zipFilePath = os.path.join(folderPath, zipFileName)
        shutil.make_archive(zipFilePath[:-4], 'zip', subFolderPath)

        # STEP 6: Convert the zip file back into a .pqt file
        newExt = '.pqt'
        directory = os.path.dirname(zipFilePath)
        fileNameWithoutExtension = os.path.splitext(os.path.basename(zipFilePath))[0]
        newFilePath = os.path.join(directory, fileNameWithoutExtension + newExt)
        shutil.move(zipFilePath, newFilePath)

        #STEP 7: Delete subFolder directory which is no longer needed
        shutil.rmtree(subFolderPath, ignore_errors=True)

        print(f"'{fileName}.pqt' has been created based on the '{datasetName}' semantic model within the Files section of your lakehouse.")

    else:
        print(f"The '{datasetName}' semantic model does not use Power Query so a Power Query Template file cannot be generated.")

create_pqt_file('IncrementalRefreshTest') #Enter semantic model name